Cancel tx JS

CraigglesO opened the pull request:
Javascript updates from #4949 

* updated jsonrpc->parity
* updated api->rpc->parity
* spelling/grammar fixes

![img1](https://cdn.pbrd.co/images/L10BJQOyo.png)
![img2](https://cdn.pbrd.co/images/L11HBuYDK.png)
![img3](https://cdn.pbrd.co/images/L12oQHYYc.png)

jacogr commented on npm-debug.log:
```diff
@@ -0,0 +1,24 @@
+0 info it worked if it ends with ok
```
Please remove.

jacogr commented:
Remove `npm-debug.log` (should probably be added to `.gitignore`), update with latest master (to pull in the changes from @tomusdrw , currently they still show in this PR)

CraigglesO commented:
FIXED! Good eye! I added to `.gitignore` for the future.

I'll definitely refactor the code when the edit button comes in.

jacogr commented on js/src/ui/TxList/TxRow/txRow.js:
```diff
@@ -137,11 +142,46 @@ class TxRow extends Component {
+
+    if (canceled) {
+      return (
+        <div className={ styles.pending }>CANCELED</div>
```
FormattedMessage?

jacogr commented on js/src/ui/TxList/TxRow/txRow.js:
```diff
@@ -137,11 +142,46 @@ class TxRow extends Component {
+    if (!isCancelOpen) {
+      return (
+        <div className={ styles.pending }>
+          <div>SCHEDULED</div>
```
FormattedMessage?

jacogr commented on js/src/ui/TxList/TxRow/txRow.js:
```diff
@@ -137,11 +142,46 @@ class TxRow extends Component {
+          <div>SCHEDULED</div>
+          <a
+            className={ styles.cancel }
+            onClick={ () => this.setState({ isCancelOpen: true }) }
```
Would prefer not using in-line closures, aligning with the codebase. (I actually thought we had a linting rule that warns about these, weird that it wasn't picked up)

jacogr commented on js/src/ui/TxList/TxRow/txRow.js:
```diff
@@ -137,11 +142,46 @@ class TxRow extends Component {
+        <a onClick={ () => this.cancelTransaction(parity, hash) }>Cancel</a>
+        <span> | </span>
+        <a onClick={ () => this.setState({ isCancelOpen: false }) }>Nevermind</a>
+      </div>
```
FormattedMessage & inline closures.

jacogr commented on js/src/ui/TxList/txList.css:
```diff
@@ -42,10 +42,11 @@
+      text-align: center;
       line-height: 1.5em;
-      opacity: 0.5;
+      color: grey;
```
Not 100% sure we use this anywhere, would prefer to keep it where it was to keep consistent (We really need to externalise all colours and just pull in CSS variables - no duplication, consistency is maintained)

CraigglesO replied:
The problem is, if opacity is dropped, the "cancel" button will also be 50% opacity LOL. Also, all font is either black or white, 0.5 opacity for both black and white revert to grey anyways.

rphmeier commented on js/src/jsonrpc/interfaces/parity.js:
```diff
@@ -643,6 +643,23 @@ export default {
+    ],
+    returns: {
+      type: Boolean,
+      desc: 'returns `true` if the upgrade to the new release was successfully executed, `false` if not.',
```
Description needs updating

rphmeier commented on js/src/views/Signer/containers/RequestsPage/requestsPage.js:
```diff
@@ -100,7 +100,7 @@ class RequestsPage extends Component {
             <FormattedMessage
               id='signer.requestsPage.noPending'
-              defaultMessage='There are no requests requiring your confirmation.'
+              defaultMessage='There are no transactions or requests requiring your confirmation.'
```
Not sure if I 100% agree with this, although getting heavily into bikeshed territory :)
Transaction signing requests are still signer requests. "transactions or requests" implies that they aren't.

rphmeier commented on js/src/views/Signer/containers/RequestsPage/requestsPage.js:
```diff
@@ -114,7 +114,7 @@ class RequestsPage extends Component {
           <FormattedMessage
             id='signer.requestsPage.pendingTitle'
-            defaultMessage='Pending Requests'
+            defaultMessage='Pending Transactions'
```
I believe we can have more than just pending transaction signing requests available here. @tomusdrw?

CraigglesO replied:
Ah! I am super glad you saw these. I meant to point it out to be disputed. I feel Requests is simply somewhat misleading. We need better terminology, correct? Otherwise, I will drop it.

rphmeier replied:
They are requests as in "requests for something to be signed". A transaction isn't valid until it's signed. But this interface is also used for more general "sign this random bit of data" as well. I agree that the terminology is a little muddy and clarification would lead to a better UX, but we also have to be careful to remain accurate.

CraigglesO commented:
TODO:
* After editing, txRow does not fully update
* former conditions do not yet get passed into edit
* I don't yet like how time is printed

Updates:
* Added Editing
* Added Time/Blocks till submit
* fixed FormattedMessage & inline closures
* fixed description


![alt image](https://cdn.pbrd.co/images/LuhoT6UDK.png)

tomusdrw commented on js/src/views/Signer/containers/RequestsPage/requestsPage.js:
```diff
@@ -100,7 +103,7 @@ class RequestsPage extends Component {
             <FormattedMessage
               id='signer.requestsPage.noPending'
-              defaultMessage='There are no requests requiring your confirmation.'
+              defaultMessage='There are no transactions requiring your confirmation.'
```
Your password in signer might be required for couple of things (as Rob mentioned):
1. Sign transaction `eth_sendTransaction`
2. Sign transaction for offline use `eth_signTransaction`
3. Sign aribtrary data `eth_sign`
4. Decrypt arbitrary data `parity_decryptMessages`

CraigglesO replied:
Gah, this one is tricky. I would appreciate thoughts of what to change it to.

To me, "requests" makes absolutely no sense from an economics standpoint.
When I first downloaded this app, I was initially thinking, "I have to approve people sending me money? That's silly". But now I know what it does.

How about something akin to,
"There is nothing requiring your confirmation."

Simple. Really lacks specificity, however.

BETTER YET, hide the freakin div all together. Say nothing.

tomusdrw replied:
I like the simple approach or we can use the opportunity to explain why there is a Signer in the first place :)

tomusdrw commented on js/src/ui/TxList/TxRow/txRow.js:
```diff
@@ -137,11 +147,98 @@ class TxRow extends Component {
+            id='ui.txList.txRow.verify.cancelEdit'
+            defaultMessage='{ which }'
+            values={ {
+              which: `${(isCancelOpen) ? 'Cancel' : 'Edit'}`
```
We should get a big red/yellow warning for transaction cancellation since it's not reliable as mentioned in RPC docs.

There are two separate cases:
1. Scheduled transactions and transactions that were not propagated to other peers - such transactions can be safely canceled, it will be removed from our local queue and never propagated to the network so it's fine.
2. Other transactions - those transactions cannot be simply "canceled" since they are already floating around a network and there is a very high chance that other peers have accepted the transaction and are relaying it further.
For those transactions we should display a warning (but still allow that) and offer another "cancelation" mechanism decribed here: https://github.com/paritytech/parity/issues/4519, so: Increase gas price, remove data, lower gas and submit transaction again.

When editing transaction we should also differentiate those two cases. (1) is safe to edit, but (2) should always increase the gas price when edited (it increases the chance of actually replacing the transaction that was already propagated).

gavofyork replied:
(2) should also zero-out the `value` and, for sanity, replace `to` with `0x000...000`.

jacogr replied:
These should be FormattedMessage as well (Should also be translatable)

CraigglesO commented on js/src/ui/TxList/TxRow/txRow.js:
```diff
@@ -165,6 +262,61 @@ class TxRow extends Component {
+    const { parity } = this.context.api;
+    const { hash, gas, gasPrice, to, from, value, input, condition } = this.props.tx;
+
+    parity.removeTransaction(hash);
```
I would consider the fact that you can't use the same gas price as a bug. In this case, you can still use the same gas price, which is automated anyways. If the user didn't have the choice for gas price, but rather the GP was automated, than I would just GP++ and purely edit.

CraigglesO commented:
TODO:
* Update store.spec.js
* Improve parity.removeTransaction description
* move two functions to store

CraigglesO commented:
`js/src/ui/TxList/store.js` Recieved a complete rewrite.
1) It was subscribing a second time to the api, but the requestsPage store was already handling that so it was pushing a new state 2x per update.
2) It did not allow for the removal of hashes.
3) Sometimes the component would update too fast and not all the promises were fulfilled so there were errors.
4) It was way more complex than it needed to be.

I still think signer store should eventually be subscribed to the api in a manner rather than update every 1.5 seconds... but I don't think it's worth posting an issue about currently.

Lemme know what you think I should edit.

tomusdrw commented on js/src/api/rpc/parity/parity.js:
```diff
@@ -395,6 +395,11 @@ export default class Parity {
       .execute('parity_removeReservedPeer', encode);
   }
 
+  removeTransaction (transactionId) {
```
`transactionHash` would be more adequate

tomusdrw commented on js/src/jsonrpc/interfaces/parity.js:
```diff
@@ -643,6 +643,34 @@ export default {
+    ],
+    returns: {
+      type: Object,
+      desc: 'the removed transaction if cancellation was successful, an error if not.',
```
note: it's rather "the removed transaction if found, an error if not" (no guarantees about success state here as well)

tomusdrw commented on js/src/ui/TxList/store.js:
```diff
@@ -14,150 +14,121 @@
+  @action addHash = (hash) => {
+    if (!this.sortedHashes.includes(hash)) {
+      this.sortedHashes.push(hash);
+      this.sortHashes();
```
Could use a sorted data structure (some kind of `SortedList`) for greater efficiency. But it's most probably not on the hot path.

CraigglesO replied:
Instead of arrays? V8 arrays used in Chrome are faster. I don't know how it is for FF or whatever.

tomusdrw replied:
Arrays are fast, but instead of sorting in `O(n log(n))` each time you instert an element you can just insert with `O(log(n))` complexity and for large number of elements this will matter.

But nevermind, now I see this is just used to store local transactions so this will be 25 transactions at most afair so it's not even worth thinking about :)

tomusdrw commented on js/src/ui/TxList/store.js:
```diff
@@ -14,150 +14,121 @@
-
-      this._pendingHashes = this.sortedHashes.filter((hash) => this.transactions[hash].blockNumber.eq(0));
-    });
+  @action removeHash = (hashes) => {
```
I think it should be rather called `setHashes` ?

CraigglesO replied:
No, this literally removes the hash from the list.

Remember above there is "addHash"

Ex. I click cancel, so the whole tx disappears from the list... This function ensures the tx disappears and the state updates.

tomusdrw replied:
Ok, I think I get the context in which this function is used, but `removeHash` seems like a function that should do the invert of `addHash` (given hash add it to the list if not present) so it should do: "given hash remove it from the list if present". This function is actually just overriding all hashes with new ones that's why I though it should be named `setHashes` instead of `removeHash`.

tomusdrw commented on js/src/ui/TxList/store.js:
```diff
@@ -14,150 +14,121 @@
+            if (tx.blockHash) {
+              eth.getBlockByNumber(tx.blockNumber)
+                .then((block) => {
+                  this.blocks[tx.blockNumber] = block;
```
This structure will be leaking memory, probably `removeHash/setHashes` should update `this.blocks` as well and remove old entries.

tomusdrw replied:
On a second glance: `this.transactions` will be leaking as well, we're never removing any entries from it.

CraigglesO replied:
Remember that react components have lifetimes. This is a Mobx store, so it dies with the component. Like you said earlier, 25ish tops are going to exist. So as soon as the component dies, so does all the data. No reason to add more code.

tomusdrw replied:
Yes, but if I leave my page open on transaction list it will consume all memory after some time. IMHO its an obvious memory leak and we shouldn't tell our users to just refresh the page from time to time.

tomusdrw commented on js/src/ui/TxList/store.js:
```diff
@@ -14,150 +14,121 @@
-      return;
-    }
+    parity.removeTransaction(hash)
+    .then(() => {
```
Indent.

jacogr commented on js/src/jsonrpc/interfaces/parity.js:
```diff
@@ -1622,7 +1650,14 @@ export default {
           from: '0xb60e8dd61c5d32be8058bb8eb970870f07233155',
           to: '0xd46e8dd67c5d32be8058bb8eb970870f07244567',
-          value: fromDecimal(2441406250)
+          gas: fromDecimal(30400),
```
Would just remove the `fromDecimal` wrappers here, i.e. the response should rather be what we receive via the actual RPC, raw.

jacogr commented on js/src/jsonrpc/interfaces/parity.js:
```diff
@@ -1622,7 +1650,14 @@ export default {
+          data: '0xd46e8dd67c5d32be8d46e8dd67c5d32be8058bb8eb970870f072445675058bb8eb970870f072445675',
+          condition: {
+            block: 354221,
+            time: new Date()
```
Same with these.

tomusdrw commented on js/src/ui/TxList/store.js:
```diff
@@ -14,150 +14,121 @@
+    const { parity } = this._api;
+    const { hash, gas, gasPrice, to, from, value, input, condition } = tx;
+
+    parity.removeTransaction(hash)
```
Why not:
```js
parity.removeTransaction(hash)
  .then(() => parity.postTransaction({ ... }))
  .then(() => txComponent.setState({editing: true})) // just swap with line above if the order should be different.
  .catch(message => this._newError({ message }));
```

CraigglesO replied:
Absolutely, I was originally going to give two different error reports with more information depending on where it failed, but I couldn't think of a way to purposely create a failure so.... LOL.

jacogr commented on js/src/ui/TxList/TxRow/txRow.js:
```diff
@@ -137,11 +149,113 @@ class TxRow extends Component {
+    if (!isCancelOpen && !isEditOpen) {
+      const pendingStatus = this.getCondition();
+
+      if (pendingStatus === 'submitting...') {
```
Not sure I'm 100% on-board with the `submitting...` - would rather make the state checking against a defined constant (e.g. `const STATUS_PENDING = 'submitting'`) and not have the `...` in there

jacogr commented on js/src/ui/TxList/TxRow/txRow.js:
```diff
@@ -137,11 +149,113 @@ class TxRow extends Component {
+        <div className={ styles.pending }>
+          <FormattedMessage
+            id='ui.txList.txRow.editing'
+            defaultMessage='EDITING'
```
It probably would make more sense having the `defaultMessage` as `editing/Editing` and using text-transform to make it uppercase. (That way we don't need to edit the text and get translations to update when we change the way we want the display the look)

(Same for `SUBMITTING` & `CANCEL`)

tomusdrw commented on js/src/views/Signer/containers/RequestsPage/requestsPage.js:
```diff
@@ -85,7 +88,9 @@ class RequestsPage extends Component {
           address=''
+          blockNumber={ blockNumber }
           hashes={ localHashes }
+          newError={ newError }
```
Isn't there an agreement that fired events should be always prefixed with `on*`? Imho thinking in terms of `props` and `events` helps to understand components architecture better (without imperative: "`createNewErrorNow` stuff)

CraigglesO replied:
I kept it like this to avoid confusion. That's a redux function I am passing in. I think it's more obvious to keep the name.

tomusdrw replied:
Ok, it's not inline with some other functions in the same file though, we map redux actions to component events like this:
```js
onConfirm={ startConfirmRequest }
onReject={ startRejectRequest }
```

jacogr commented on js/src/ui/TxList/TxRow/txRow.js:
```diff
@@ -137,11 +149,113 @@ class TxRow extends Component {
+          />
+        </a>
+        <span>{' | '}</span>
+        <a onClick={ this.nevermind }>
```
Would name this something else, e.g. not `nevermind` (`revertEdit` or `cancelEdit` or ...)

jacogr commented on js/src/ui/TxList/TxRow/txRow.js:
```diff
@@ -165,6 +279,48 @@ class TxRow extends Component {
+      }
+    } else if (blockNumber) {
+      block = blockNumber.minus(block);
+      return (block.toNumber() < 0)
```
FormattedMessage here

CraigglesO replied:
All string returns pulled from `getCondition` as seen here are formatted on line `205`

tomusdrw commented on js/src/views/Signer/store.js:
```diff
@@ -95,7 +95,11 @@ export default class SignerStore {
-        this.setLocalHashes(Object.keys(localTransactions));
+        const keys = Object
+          .keys(localTransactions)
+          .filter((key) => localTransactions[key].status !== 'invalid');
```
Why do we skip `invalid` transactions here? I think it would be quite usefull to re-submit transaction cancelled by mistake. Maybe just change the way those are displayed?

CraigglesO replied:
Huh. I don't know how else to remove canceled transactions. When removeTransaction is called, it does not actually remove the tx from localTransactions, but rather gives 'invalid', so editing and cancelling tx, they don't go away...

Like we discussed before. To avoid having to change the gasPrice, I completely cancel the TX and then resubmit. It works beautifully. If you hit cancel, the submittal bar pops up with every single detail from the just cancelled tx still there. So you just can sign and resubmit for the exact tx back again.

tomusdrw replied:
Maybe keep them but display them in UI differently? Actually I will prepare a special `status` for canceled transactions. `invalid` can happen in some other cases as well and the whole `localTransactions` thing was created to never loose any of transactions that were sent from local node.

jacogr commented on js/src/ui/TxList/store.js:
```diff
@@ -14,150 +14,121 @@
-    if (!blockNumbers || !blockNumbers.length) {
-      return;
-    }
+    parity.removeTransaction(hash)
```
Slight style grumble, would prefer `.then` & `.catch` indented, i.e.

```
parity
  .removeTransaction(...)
  .then(...)
  .catch(...)
```

jacogr commented on js/src/ui/TxList/store.js:
```diff
@@ -14,150 +14,121 @@
+
+    parity.removeTransaction(hash)
+      .then(() => {
+        parity.postTransaction({
```
Indentation of this block and the containing one. In this case (and I generally do exactly what you've done myself), it could aid readability to have `transaction` as a `const` and passing it in.

jacogr commented on js/src/ui/TxList/txList.css:
```diff
@@ -42,10 +42,11 @@
+      text-align: center;
       line-height: 1.5em;
-      opacity: 0.5;
+      color: grey;
```
Would keep the original to align with the rest of the UI. (Close enough, but not 100% the same in terms of colors here, we don't use gray elsewhere.) Yes... we need to externalize all color and related vars and just pull them in.

CraigglesO replied:
All links and others (sumbitting/ cancel/ edit ...) will be 50% as well. Which is a problem to see/links look strange. I don't mind adding more css, so that the main is 0.5 opacity. But I can't keep the css the way it is. Suggestions are welcome.

jacogr replied:
Ahh, ok, the links as well, makes sense. Maybe be more specific then with `>span` (FormattedMessage)

jacogr commented on js/src/ui/TxList/txList.js:
```diff
@@ -35,19 +35,16 @@ class TxList extends Component {
-  store = new Store(this.context.api);
-
   componentWillMount () {
+    this.store = new Store(this.context.api, this.props.newError);
```
Any reason for moving this? (They are equivalent)

CraigglesO replied:
newError won't get loaded in until CWM due to lifetimes. I should really remove:
`this.store.loadTransactions(this.props.hashes);` and have that part of the constructor method too....

tomusdrw replied:
I think you should also cater for `newError` being changed during the lifetime of the component (or when component will be re-used with different props). Isn't it possible to avoid passing the error directly?
Like:
```js
store: new Store(this.context.api, (...args) => this.props.newError(...args))
// or
//store: new Store(this.context.api, (...args) => this.props.newError ? this.props.newError(...args) : null)
```

CraigglesO replied:
newError is a redux action. So I don't really know what you mean. Basically it goes like this:

action->reducer->store->component.

passing in an argument to newError is the action phase.

tomusdrw replied:
Ok, so the issue was that this:
```js
store: new Store(this.context.api, this.props.newError)
```
doesn't work cause `this.props.newError` is not available yet because of the lifetime issues you mentioned. So your solution was to do:
```js
componentWillMount () {
  this.store = new Store(this.context.api, this.props.newError);
}
```
And that works correctly. My point is that it's still not working for all cases (it's not generic enough), cause `this.props.newError` can change if the parent component decides to change it. So proper implementation should look like this:
```js
componentWillMount () {
  this.store = new Store(this.context.api, this.props.newError);
}
componentWillReceiveProps (newProps) {
  if (this.props.newErrror !== newProps.newError) {
    this.store = new Store(this.context.api, newProps.newError);
    // + copy state of the previous store most probably.
  }
}
```
but that's way to complicated. So instead I proposed a simpler solution:
```js
store: new Store(this.context.api, (...args) => this.props.newError(...args))
```

So store has a reference to a lambda function that will dynamically dispatch to `this.props.newError` so even if `this.props.newError` is changed the Store doesn't need to be updated.

IMHO even if `this.props.newError` is not subject to change currently we shouldn't rely on that cause in future it may and such code may cause a wierd errors in the future that are easy to avoid now.

Hope this explains my motivations and solution ;)

ngotchac replied:
So I think there's a misunderstanding here. The `newError` prop is actually added in the `connect` part of the Component, managed by `react-redux`. This just adds the `newError` function prop from the imported function in `./redux/.../errorActions` or something similar. It is not passed from a parent (or at least shouldn't be). As it's an import, the function won't change during the lifetime of the application.

tomusdrw commented on js/src/ui/TxList/store.js:
```diff
@@ -14,150 +14,121 @@
-
-            if (tx.blockNumber && tx.blockNumber.gt(0)) {
-              blockNumbers.push(tx.blockNumber.toNumber());
+    _txhashes
```
Imagine this:
```js
this.sortedHashes = ['A', 'B']
_txhashes = ['A', 'C']
```
It has the same length, but it's not the same array so it will fall into this branch and you'll end up with
```
this.sortedHashes = ['A', 'B', 'C' ]
```
I believe this could be a case when for instance `B` is canceled in favour of `C` and we'll fail to register cancelation first.

Actually it does way more work then required, cause it fetches `eth.getTransactionByHash` for each transaction even if those transactions are already in `this.transactions`.

CraigglesO replied:
I couldn't think at the time of a case where in the ui this could happen. 

But I believe you if you think this will be a problem in the future, so I will change it.

edit - If anything changes, I resubmit for all tx just incase there were any changes along the way. I was worried if I filtered the old tx there wouldn't be changes (future edits)

CraigglesO replied:
Relooking at my code now, this is actually not an issue.
this.sameHashList checks the equality of the two (this.containsAll both directions),
so:
(['A','B'], ['B','A']) === true
(['A','B'], ['A','C']) === false

It will be a two step process... if the scenario you explained happened, the first thing it would do is add 'C' to this.sortedHashes, than it would see that it has more variables than its counterpart and remove B.

tomusdrw replied:
Yes, only if on the second step this condition would be satisfied: `_txhashes.length < this.sortedHashes.length`.
I suppose it will reach eventualy consistency, but the fact is that it's not working 100% correctly in all cases and may cause some heisenbugs.

CraigglesO commented:
- [ ] removeHash => setHashes
- [ ] update parity.removeTransaction details
- [ ] transactionId => transactionHash (js/src/api/rpc/parity/parity.js)
- [ ] formatMessages
- [ ] never mind => something better
- [ ] text-transform => uppercase
- [ ] submitting… => Submitting
- [ ] remove fromDecimal wrappers (js/src/jsonrpc/interfaces/parity.js)
- [ ] newError => onNewError
- [ ] filter invalid => filter canceled
- [ ] update then catch formatting

CraigglesO commented:
Instead of bulk handling change, this version will handle individual changes noticed.

I also stumbled upon hashes that have null returned via `eth.getTransactionByHash`. So cleaned that up too.

If there are any other minute details you guys see, let me know.

tomusdrw commented on js/src/jsonrpc/interfaces/parity.js:
```diff
@@ -1674,9 +1674,9 @@ export default {
-          gas: fromDecimal(30400),
-          gasPrice: fromDecimal(10000000000000),
-          value: fromDecimal(2441406250),
+          gas: 30400,
```
Just curious, why was it changed? It won't work with decimals on RPC layer.

CraigglesO replied:
Jaco wanted them removed. If you are sure they are required than absolutely I will put them back.

tomusdrw replied:
I think he meant to convert the decimals into hex values so that the example is not using any functions (but I'm unsure though).

tomusdrw commented on js/src/ui/TxList/store.js:
```diff
@@ -14,150 +14,131 @@
+  @action removeHash = (hash) => {
+    this.sortedHashes.remove(hash);
+    delete this.transactions[hash];
+    delete this.blocks[hash];
```
This should be:
```
let tx = this.transactions[hash];
if (tx) {
  delete this.transactions[hash];
  delete this.blocks[tx.blockNumber];
}
```
looks good otherwise.

CraigglesO commented:
This should do it. 👍

jacogr commented on js/src/ui/TxList/TxRow/txRow.js:
```diff
@@ -165,6 +282,48 @@ class TxRow extends Component {
+      block = blockNumber.minus(block);
+      return (block.toNumber() < 0)
+        ? block.abs().toFormat(0) + ' blocks left'
+        : 'submitting';
```
Just make the strings here translatable. (Also line 247 is still, as part of the value for `which`, non-translatable)

jacogr commented:
Couple of last i18n issues, but apart from that looking good.

CraigglesO commented:
Let me know 💃

jacogr commented:
One last conflict.

CraigglesO commented:
DONE.

5chdn commented:
Can we have this backported to beta? It's causing a support nightmare :)
